### PR TITLE
Lots of updates

### PR DESCRIPTION
**Pull request by scottpchow23**
### Yay for GitHub API Integration!

Among other things:

* Stubbing for octokit and http requests in general with use of webmock

* Created classes now verify that the Machine User was added as an owner to the organization.

* Machine User invites users joining a class to the class organization.

* Header checkbox bug was fixed.

### CODE REVIEW

File: app/models/course.rb [hunk 1]
```diff
@@ -41,22 +51,24 @@ def users
   def import_students(file, header_map, header_row_exists)
     ext = File.extname(file.original_filename)
     spreadsheet = Roo::Spreadsheet.open(file, extension: ext)
-    
+
     # get index for each param
-    id_index = header_map.index("student_id")
+    id_index = header_map.index("perm")
```
Comment (garethgeorge) [inline app/models/course.rb]: renaming student_id to perm is fine, though less general to other schools.

File: app/models/course.rb [hunk 2]
```diff
@@ -98,4 +110,23 @@ def export_students_to_csv
...
+
+  def delete_roster_students(spreadsheet, header_row_exists, perm_index)
+    perms = sort_perms_into_hash(spreadsheet, header_row_exists, perm_index)
+    self.roster_students.each do |student|
+      perm = perms[student.perm]
+      if perm.blank?
+        student.destroy
+      end
+    end
+  end
```
Comment (garethgeorge) [inline app/models/course.rb]: I wonder if this function name is misleading / why this got its own function.

This appears to be filtering a roster for invalid students? Or perhaps deduplicating.

Would a better name be 'deduplicate_roster_students'

Or sanitize_roster_students

File: app/models/course.rb [hunk 3]
```diff
@@ -98,4 +110,23 @@ def export_students_to_csv
...
+
+  def sort_perms_into_hash(spreadsheet, header_row_exists, perm_index)
```
Comment (garethgeorge) [inline app/models/course.rb]: both of these methods should probably be moved into a private section if they aren't already.

File: app/views/layouts/_messages.html.erb [hunk 4]
```diff
@@ -3,7 +3,7 @@
   <% if msg.is_a?(String) %>
     <div class="alert alert-dismissible alert-<%= name.to_s == 'notice' ? 'success' : 'danger' %>">
       <button type="button" class="close" data-dismiss="alert" aria-hidden="true">&times;</button>
-      <%= content_tag :div, msg, :id => "flash_#{name}" %>
+      <%= content_tag :div, sanitize(msg), :id => "flash_#{name}" %>
```
Comment (garethgeorge) [inline app/views/layouts/_messages.html.erb]: is this to prevent HTML injection?

File: db/schema.rb [hunk 5]
```diff
@@ -22,7 +25,7 @@
   create_table "roles", force: :cascade do |t|
     t.string "name"
     t.string "resource_type"
-    t.integer "resource_id"
+    t.bigint "resource_id"
```
Comment (garethgeorge) [inline db/schema.rb]: Why did this have to be changed?

File: db/schema.rb [hunk 6]
```diff
@@ -37,8 +40,8 @@
     t.string "email"
     t.datetime "created_at", null: false
     t.datetime "updated_at", null: false
-    t.integer "course_id"
-    t.integer "user_id"
+    t.bigint "course_id"
+    t.bigint "user_id"
```
Comment (garethgeorge) [inline db/schema.rb]: again wondering what issue this fixed... is there some library assuming that these are bigints or are we worried about having more than 2^32 courses? :)

File: db/schema.rb [hunk 7]
```diff
@@ -67,11 +70,13 @@
...
-    t.integer "role_id"
+    t.bigint "user_id"
+    t.bigint "role_id"
     t.index ["role_id"], name: "index_users_roles_on_role_id"
     t.index ["user_id", "role_id"], name: "index_users_roles_on_user_id_and_role_id"
     t.index ["user_id"], name: "index_users_roles_on_user_id"
   end
 
+  add_foreign_key "roster_students", "courses"
+  add_foreign_key "roster_students", "users"
```
Comment (garethgeorge) [inline db/schema.rb]: good.

**Comment by garethgeorge** [comment]
@wpeery @scottpchow23 @shudaniel

**Comment by garethgeorge** [comment]
Looking like great progress!

File: db/schema.rb (hunk 6 above)
Comment (wpeery) [reply db/schema.rb]: This was changed accidentally, by something I was playing around with. Im not sure what exactly caused this to happen but ill change it back

File: app/views/layouts/_messages.html.erb (hunk 4 above)
Comment (scottpchow23) [reply app/views/layouts/_messages.html.erb]: Yeah it was more force of habit. We're creating a link from the organization name that leads to the invite, and even though GitHub org names should be sanitized to begin with, it doesn't hurt to sanitize this.

File: app/controllers/courses_controller.rb [hunk 8]
```diff
@@ -93,4 +94,8 @@ def set_course
     def course_params
       params.require(:course).permit(:name,:course_organization)
     end
+
+    def add_instructor
```
[the remark this answers is not in the document]
Comment (wpeery) [reply app/controllers/courses_controller.rb]: Good point will fix

**Comment by shudaniel** [comment]
Conrad said to just merge so we can test.

**Pull request merged**